modules/storage-rocksdb/src/main/java/org/apache/ignite/internal/storage/rocksdb/RocksDbMvPartitionStorage.java

```diff
@@ -0,0 +1,494 @@
+
+        try {
+            // Read a value associated with pending write.
+            byte[] valueBytes = db.get(cf, keyBuf.array(), 0, keyBuf.position());
+
+            // Delete pending write.
+            db.delete(cf, writeOpts, keyBuf.array(), 0, keyBuf.position());
+
+            // Add timestamp to the key put the value back into the storage.
+            putTimestamp(keyBuf, timestamp);
+
+            db.put(cf, writeOpts, keyBuf.array(), 0, keyBuf.position(), valueBytes, 0, valueBytes.length);
+        } catch (RocksDBException e) {
+            throw new StorageException(e);
+        }
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    public @Nullable BinaryRow read(IgniteRowId rowId, @Nullable Timestamp timestamp) {
+        // Prepare a buffer with partition id and row id.
+        ByteBuffer keyBuf = prepareHeapKeyBuf(rowId);
+
+        try (
+                // Set next partition as an upper bound.
+                var readOpts = new ReadOptions().setIterateUpperBound(upperBound);
+                RocksIterator it = db.newIterator(cf, readOpts)
+        ) {
+            if (timestamp == null) {
+                // Seek to the first appearance of row id if timestamp isn't set.
+                // Since timestamps are sorted from newest to oldest, first occurance will always be the latest version.
+                // Unfortunately, copy here is unavoidable with current API.
+                it.seek(Arrays.copyOf(keyBuf.array(), keyBuf.position()));
+            } else {
+                // Put timestamp restriction according to N2O timestamps order.
+                putTimestamp(keyBuf, timestamp);
+
+                // This seek will either find a key with timestamp that's less or equal than required value, or a different key whatsoever.
+                // It is guaranteed by descending order of timestamps.
+                it.seek(keyBuf.array());
+            }
+
+            // Return null if nothing was found.
+            if (invalid(it)) {
+                return null;
+            }
+
+            // There's no guarantee that required key even exists. If it doesn't, then "seek" will point to a different key, obviously.
+            // To avoid returning its value, we have to check that actual key matches what we need.
+            // Here we prepare direct buffer to read key without timestamp. Shared direct buffer is used to avoid extra memory allocations.
+            ByteBuffer directBuffer = MV_KEY_BUFFER.get().position(0).limit(ROW_ID_OFFSET + igniteRowIdSize);
+
+            it.key(directBuffer);
+
+            // Comparison starts from the position of the row id.
+            directBuffer.position(ROW_ID_OFFSET);
+
+            // Return null if seek found a wrong key.
+            if (rowId.compareTo(directBuffer, false) != 0) {
+                return null;
+            }
+
+            // Get binary row from the iterator. It has the exact payload that we need.
```

Comment:
Too many comments, to my taste. This line seems obvious :)

Reply:
My idea was - the more comments, the better. This adds consistency even if comment is obvious - method is split into small digestible sections